src/include/miopen/solver/implicitgemm_ck_util.hpp

```diff
@@ -157,5 +160,406 @@ ConvSolution InitAnyInvokerFactory(const ProblemDescriptionType& problem,
+        index     = idx;
+        assert(index >= 0);
+        buf_offset = wt.GetOffset(index);
+    }
+
+    void AllocBuffer(const Handle& handle, Data_t workSpace)
+    {
+        buf_handle = handle.CreateSubBuffer(workSpace, buf_offset, tensor_sz);
+        assert(buf_handle.get());
+    }
+
+    Data_t GetBufferPtr() const { return buf_handle.get(); }
+
+    void ConvertFrom(const Handle& handle, const std::vector<Kernel>& kernels, ConstData_t in_ptr)
+    {
+        Run(handle, kernels, buf_handle.get(), in_ptr);
+    }
+
+    void ConvertTo(const Handle& handle, const std::vector<Kernel>& kernels, Data_t out_ptr)
+    {
+        Run(handle, kernels, out_ptr, buf_handle.get());
+    }
+
+    TransposeInstance()                         = delete;
+    TransposeInstance(const TransposeInstance&) = default;
+    TransposeInstance(TransposeInstance&&)      = default;
+    ~TransposeInstance()                        = default;
+
+private:
+    void Run(const Handle& handle,
+             const std::vector<Kernel>& kernels,
+             Data_t out_ptr,
+             ConstData_t in_ptr)
+    {
+        assert(out_ptr);
+        assert(in_ptr);
+        assert(kernels.size() > index);
+
+        kern_args[0] = out_ptr;
+        kern_args[1] = in_ptr;
+
+        handle.Run(kernels[index])(kern_args);
+        if(handle.IsProfilingEnabled())
+        {
+            handle.AccumKernelTime(handle.GetKernelTime());
+        }
+    }
+};
+
+template <typename InTransSolType, typename WeiTransSolType, typename OutTransSolType>
+struct TransposeInstanceMaker2D
+{
+
+    template <typename CKArgsType>
+    std::tuple<TransposeInstance, TransposeInstance, TransposeInstance>
+    operator()(ConvSolution& result,
+               const ExecutionContext& ctx,
+               const ProblemDescription& problem,
+               const CKArgsType& ck_args) const
+    {
+
+        InTransSolType tr_in(
+            ctx, problem.GetInDataType(), ck_args.N, ck_args.C1, ck_args.Hi, ck_args.Wi);
+
+        WeiTransSolType tr_wei(
+            ctx, problem.GetWeightsDataType(), ck_args.K1, ck_args.C, ck_args.Y, ck_args.X);
+
+        OutTransSolType tr_out(
+            ctx, problem.GetOutDataType(), ck_args.N, ck_args.K1, ck_args.Ho, ck_args.Wo);
+
+        result.construction_params.insert(
+            result.construction_params.end(),
+            {tr_in.GetKernelInfo(), tr_wei.GetKernelInfo(), tr_out.GetKernelInfo()});
+
+        constexpr size_t buf_alignment = 256;
+        MultiBufferWorkspaceTraits wt({tr_in.GetOutputTensorSize(),
+                                       tr_wei.GetOutputTensorSize(),
+                                       tr_out.GetOutputTensorSize()},
+                                      buf_alignment);
+
+        return std::make_tuple(TransposeInstance(tr_in, 0, wt),
+                               TransposeInstance(tr_wei, 1, wt),
+                               TransposeInstance(tr_out, 2, wt));
+    }
```

Comment:
A little bit of C++
You can add such function
```CPP
template <typename I, typename W, typename O>
std::tuple<TransposeInstance, TransposeInstance, TransposeInstance> tuple_cvt(ConvSolution& result, std::tuple<I, W, O>&& t)
{
    return std::apply([&result](auto&&... args){
        result.construction_params.insert(result.construction_params.end(), {args.GetKernelInfo()...});

        constexpr size_t buf_alignment = 256;
        MultiBufferWorkspaceTraits wt({args.GetOutputTensorSize()...}, buf_alignment);
        size_t n = 0;
        return std::tuple{TransposeInstance(args, n++, wt)...};
    }, t);
}
```
and replace this operator with following code both for 2d and 3d cases (adjusting ck_args sequence accordingly)

```suggestion
    template <typename CKArgsType>
    std::tuple<TransposeInstance, TransposeInstance, TransposeInstance>
    operator()(ConvSolution& result,
               const ExecutionContext& ctx,
               const ProblemDescription& problem,
               const CKArgsType& ck_args) const
    {
        return tuple_cvt(result,
                  InTransSolType(ctx, problem.GetInDataType(), ck_args.N, ck_args.C1, ck_args.Hi, ck_args.Wi),
                  WeiTransSolType(ctx, problem.GetWeightsDataType(), ck_args.K1, ck_args.C, ck_args.Y, ck_args.X),
                  OutTransSolType(ctx, problem.GetOutDataType(), ck_args.N, ck_args.K1, ck_args.Ho, ck_args.Wo));
    }
```

Reply:
You can even get rid of `TransposeInstanceMaker2D` and `TransposeInstanceMaker3D` structs, replace it with modified `tuple_cvt` and localize everything inside `ConvSolution InitInvokerFactoryFwdNCHW` function with `if constexpr`. 

It can minimize lots of entities and dependencies, make the code and the meanings more localized and reduce amount of code as well.

Reply:
Many good suggestions. Let me see how I can simplify the code. My end goals are to keep the code generic and simple. So if it's not the most minimal code, I'm OK with it. I don't want to replace more lines with less lines of more complex code. That doesn't serve C++ codebases in the long-term as people of different skill-levels join the project, and most complex code means difficulty to understand the code. 
CC: @JehandadKhan.

Reply:
A bit simplified, but more copy-pasted version.
```CPP
template <typename I, typename W, typename O>
std::tuple<TransposeInstance, TransposeInstance, TransposeInstance> tuple_cvt(ConvSolution& result, I&& i, W&& w, O&& o)
{
        result.construction_params.insert(result.construction_params.end(), {i.GetKernelInfo(), w.GetKernelInfo(), o.GetKernelInfo()});

        constexpr size_t buf_alignment = 256;
        MultiBufferWorkspaceTraits wt({i.GetOutputTensorSize(), w.GetOutputTensorSize(), o.GetOutputTensorSize()}, buf_alignment);
        return {TransposeInstance(i, 0, wt), TransposeInstance(w, 1, wt), TransposeInstance(o, 2, wt)};
}
```